### PR TITLE
feat: interest-2.0

### PR DESCRIPTION
**Pull request by Teolhyn**
- Moved interest rate calculation to pools -> fits better this architecture and also ends up with less x-contract calls.
- Added `Accrual` and `AccrualLastUpdated` parameter for pools
- Changed interest calculations time resolution from ledger sequence to seconds.
- Changed deprecated register_contract to register.
- Modified loan interest calculations to use accrual.
- Added accrual interest calls to all pool functions
- Fixed a potential divide-by-zero error
- Added add_interest call to basic loan functions
- Because of the previous point, repay couldn't close loans so I added repay_and_close function that always pays the whole loan. This has to be implemented to the front-end.
- Updated Stellar-SDK

### CODE REVIEW

File: contracts/loan_manager/src/contract.rs [hunk 1]
```diff
@@ -183,94 +182,55 @@ impl LoanManager {
...
+        let Loan {
+            borrower,
+            borrowed_from,
+            collateral_amount,
+            borrowed_amount,
+            collateral_from,
+            #[allow(unused_variables)]
+            health_factor,
+            unpaid_interest,
+            last_accrual,
```
Comment (kovipu) [inline contracts/loan_manager/src/contract.rs:196]: If you don't use all fields, you can do `..`

https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-remaining-parts-of-a-value-with-

File: contracts/loan_manager/src/contract.rs [hunk 2]
```diff
@@ -183,94 +182,55 @@ impl LoanManager {
...
+        } = Self::get_loan(e, user.clone());
 
-        let current_ledger = e.ledger().sequence();
+        let borrow_pool_client = loan_pool::Client::new(e, &borrowed_from);
+        let collateral_pool_client = loan_pool::Client::new(e, &collateral_from);
 
-        let key: LoansDataKey = LoansDataKey::LastUpdated;
-        let previous_ledger: u32 = e.storage().persistent().get(&key).unwrap_or(current_ledger); // If there is no previous ledger, use current.
+        let token_ticker = borrow_pool_client.get_currency().ticker;
+        let token_collateral_ticker = collateral_pool_client.get_currency().ticker;
```
Comment (kovipu) [inline contracts/loan_manager/src/contract.rs:203]: You could use unstructuring here as well, might clean the code a tiny bit. Up to you.

File: contracts/loan_pool/src/contract.rs [hunk 3]
```diff
@@ -165,6 +178,29 @@ impl LoanPoolContract {
...
     }
 
+    pub fn add_interest_to_accrual(e: Env) {
+        const DECIMAL: i128 = 10000000;
+        const SECONDS_IN_YEAR: u64 = 31_556_926;
+
+        let current_timestamp = e.ledger().timestamp();
+        let accrual = pool::read_accrual(&e);
+        let accrual_last_update = pool::read_accrual_last_updated(&e);
+        let ledgers_since_update = current_timestamp - accrual_last_update;
```
Comment (kovipu) [inline contracts/loan_pool/src/contract.rs:188]: Due to clock synchronization issues, this value could become negative and screw things up.

Should we add `ledgers_since_update >= 0` and handle errors gracefully just to be extra safe?

Unlikely scenario, I know.

**Review by kovipu — APPROVED** [review]

File: contracts/loan_pool/src/contract.rs (hunk 3 above)
Comment (Teolhyn) [reply contracts/loan_pool/src/contract.rs:188]: It's ledger timestamp so unless blockchain runs backwards it won't happen :)

**Pull request merged**